### ISSUE TITLE
WithParsed(async swallows exceptions

### ISSUE THREAD

**Issue by StijnVA** [commandlineparser/commandline#773]
### Some examples of the use of WithParsed /  WithParsedAsync with expected behavior
```
 internal static class Program
    {
        private static void Main(string[] args)
        {
            Parser.Default.ParseArguments<Options>(args).WithParsed(_ => throw new Exception());
```
=> Unhandled exception. System.Exception: Exception of type 'System.Exception' was thrown. 
As Expected.

```
 internal static class Program
    {
        private static async Task Main(string[] args)
        {
            await Parser.Default.ParseArguments<Options>(args).WithParsedAsync(_ => throw new Exception());
```
=> Unhandled exception. System.Exception: Exception of type 'System.Exception' was thrown. 
As Expected.

```
 internal static class Program
    {
        private static async Task Main(string[] args)
        {
            await Parser.Default.ParseArguments<Options>(args).WithParsedAsync(async _ =>
            {
                await Task.CompletedTask; //just for demonstration
                throw new Exception();
            });
```
=> Unhandled exception. System.Exception: Exception of type 'System.Exception' was thrown. 
As Expected.

### When using WithParsed with an async expresion in a synchrone context, it swallows the exception

```
    internal static class Program
    {
        private static void Main(string[] args)
        {
            Parser.Default.ParseArguments<Options>(args).WithParsed(async _ =>
            {
                await Task.CompletedTask; //just for demonstration
                throw new Exception();
            });
```
=> Process finished with exit code 0.
Should also notify about Unhandled exception.

**Comment by elgonzo** [comment]
No, not a bug in CommandLineParser. This is how async methods work, and that's why there is an _WithParsedAsync_ method specifically for async callbacks. What you are attempting here is wrong and a bad(TM) idea.

Your problem stems from trying to treat an async method in the same fashion as ye'olde synchronous methods. This is never going to work...

Exceptions occuring in an async method bubble "upwards" to the original caller through the `Task` object returned by the async method. This means, to properly handle exceptions occuring in an async method, the caller of the async method has to handle the `Task` object returned by that method TOO. In other words, exceptions thrown in async methods only have a chance to reliably bubble "upwards" if the caller of the async method handles the Task object.

If you find this difficult to understand, look at this little program: (dotnetfiddle https://dotnetfiddle.net/qP2ozk)
```C#
class Program
{
    static void Main(string[] args)
    {
        try
        {
            SomeAsyncMethod();
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Caught exception!!! {ex.GetType().FullName}");
            return;
        }

        Console.WriteLine("Yupp, as expected, no exception caught");
    }


    static async Task SomeAsyncMethod()
    {
        await Task.CompletedTask; //just for demonstration
        throw new Exception();
    }
}
```
If you execute this, you'll notice that no exception is being caught. Why?  Because the caller of _SomeAsyncMethod_ - the _Main_ method - does _NOT_ handle the Task object returned by _SomeAsyncMethod_ (_Main_ does not employ the _await_ keyword -- which under the hood deals with the returned `Task` object, nor does _Main_ handle the returned `Task` object explicitly).

<BR>

Now, _WithParsed_ obviously doesn't handle `Task` objects, because it is not meant to be used with async methods but rather with Action&lt;T&gt;-like delegate types.

And that's what  _WithParsedAsync_ is for -- handling the `Task` objects returned by an async callback. You will notice that the callback delegate type for _WithParsedAsync_ is different from that of _WithParsed_ - it is a Func&lt;Task&gt;-like delegate type here. You will also notice that _WithParsedAsync_ itself returns a `Task` object, again because it is an async method itself, and thus allowing any possible exception occuring to further "bubble" upwards in the call chain.

There is another related reason why there are both _WithParsed_ and _WithParsedAsync_, and not just a single _WithParsed_ .You might have already realized that _WithParsed_ will not be able to deal with async methods, unless the action-like delegate type would change into some Func&lt;Task&gt;-like delegate type. But then you have turned _WithParsed_ into a pointless copycat of _WithParsedAsync_, and -- crucially and more importantly -- lost the ability to use Action&lt;T&gt;-like delegates/callback methods.
